### PR TITLE
Invalidate cache load status on historical as soon as it is served

### PR DESCRIPTION
**Pull request by kfaraz**
### Description

Follow up to #11717 

This PR attempts to fix the behaviour in the following scenarios:

#### Case 1: Coordinator gets incorrect response in case of a quick load-drop-load
1. Historical receives HTTP request to load segment S.
2. Load succeeds and the cache now has an entry to the effect "LOAD S was SUCCESS".
3. HTTP request completes with SUCCESS but the entry is still there in the cache.
4. Historical receives request to drop segment S and request finishes successfully but cache still has entry for "LOAD S was SUCCESS"
5. Historical receives request to reload segment S. Since we already have an entry in cache, historical incorrectly responds with "LOAD S was SUCCESS" even though the segment has already been dropped from the historical.

#### Case 2: Coordinator sees a repeatedly failing load as PENDING rather than FAILED
1. Coordinator requests Historical to load segment.
2. Request completes while load is still in progress and gets back a PENDING response
3. Then load fails and FAILED status is cached
4. Coordinator checks with the Historical if it has loaded the segment.
5. Since we already have a cached FAILED status, we retry the load and send back a PENDING response instead (see snippet below)
6. Repeat

https://github.com/apache/druid/blob/efb32810c48fd9c0bf81cefb15179f74a4cde661/server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java#L562-L563

#### Case 3:
1. Coordinator requests historical to load segment.
2. Historical queues load task on executor.
3. Coordinator requests historical to drop segment and the Historical
   - unannounces the segment immediately
   - marks the drop request as a success in the cache
   - adds an entry to `segmentsToDrop`
   - schedules drop task on executor for 30s later
4. Load task starts and removes this segment from `segmentsToDrop` and proceeds with downloading the segment.
5. Load task finishes and announces the segment but is not able to update success status since the latest request is a Drop.
6. Drop task does not do anything since the segment is not present in `segmentsToDrop` anymore.
7. Thus the Drop request is considered a success but the segment files are never deleted.
8. This might cause the Coordinator to assign this Historical beyond its disk capacity

Similarly a drop that comes before a load might result in partial or complete delete of segment files but the load task might continue to think that it succeeded.

### Classes to review
- `SegmentLoadDropHandler`
- `SegmentLoadDropHandlerTest`

### Changes
- Rename `DataSegmentChangeRequestAndStatus` to `DataSegmentChangeResponse`
- Move `DataSegmentChangeResponse` out of `SegmentLoadDropHandler` as it is a POJO returned to the client and need not be tied to a specific loader implementation.
- Update `requestStatuses` cache to be a map from `DataSegment` to the full `DataSegmentChangeResponse`
- Remove entry from cache as soon as it is consumed to be sent back to the client
- Remove entry from cache if a different type of request is received for the same segment
- Rename and rearrange some methods for readability
- Clean up tests and add more test cases for the new behaviour

### Release note
Fix bug on historical when a load-drop-load of a segment happens in quick succession.

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java [hunk 1]
```diff
@@ -789,15 +789,20 @@ private CustomSettableFuture(
 
     public void resolve()
     {
-      synchronized (statusRefs) {
+      synchronized (requestStatusesLock) {
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:792]: Thanks for calling it out. Yes, it is better if I just add a comment explaining the logic.

I made the switch to make sure that the update invoked by `requestStatuses.invalidate()` is safeguarded.
The current synchronization on `statusRefs` essentially meant that only one thread can be executing the `resolve` of a given `CustomSettableFuture` at any point. That condition still holds but with less concurrency, i.e. in the new code, only one thread can be executing the `resolve` of _any_ `CustomSettableFuture` at any point.

So, in other words, while multiple segments might be loaded by the historical in parallel, the futures will only be resolved serially.

This will not have any other impact on the correctness or performance of this code.

File: server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java [hunk 2]
```diff
@@ -789,15 +789,20 @@ private CustomSettableFuture(
...
           return;
         }
 
-        List<DataSegmentChangeRequestAndStatus> result = new ArrayList<>(statusRefs.size());
-        statusRefs.forEach(
-            (request, statusRef) -> result.add(new DataSegmentChangeRequestAndStatus(request, statusRef.get()))
-        );
+        final List<DataSegmentChangeRequestAndStatus> result = new ArrayList<>(statusRefs.size());
+        statusRefs.forEach((request, statusRef) -> {
+          // Remove complete statuses from the cache
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:799]: As long as the coordinator doesn't give up on a request it has sent to the historical, we should be good.
i.e. once the coordinator has asked a historical to LOAD a segment, it should not ask the historical to DROP the same segment until the LOAD has failed or succeeded.

Otherwise, we might keep the cached value of LOAD success around and send serve it to the second load request which comes after the drop.

[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:799]: This code could probably do with some cleanup and doing what had been originally suggested [in this comment](https://github.com/apache/druid/pull/11717#discussion_r723833447) i.e. maintaining the cache (or really just a concurrent map) at the `DataSegment` level itself rather than `DataSegmentChangeRequest`.

But I wanted to keep the footprint of this PR low so that we could include it in Druid 27 release. We can take up a more holistic overhaul later. Let me know what you think.

File: server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gianm) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:792]: Thanks for the explanation. It looks like there is no longer anything synchronizing on `statusRefs`, so the new code is somewhat simpler. Please do add a comment to `requestStatusesLock`, or use `@GuardedBy` on the things it is meant to guard, as appropriate, in order to make the logic easier to follow for others.

File: server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gianm) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:799]: > once the coordinator has asked a historical to LOAD a segment, it should not ask the historical to DROP the same segment until the LOAD has failed or succeeded.

Could this happen after Coordinator leader change or restart?

File: server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:815]: Added an `@GuardedBy` for `requestStatuses` and some comments.

**Comment by kfaraz** [comment]
Also moved out `DataSegmentChangeResponse` into a class of its own.

File: server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:833]: Yes, I feel those are the only cases when this is likely to happen.

**Comment by kfaraz** [comment]
@gianm , there is another case which the current code was behaving weirdly. If a load keeps failing repeatedly on a historical, the coordinator may always get a PENDING response rather than FAILED.

Typical scenario:
- Coordinator requests Historical to load segment.
- Request completes while load is still in progress and gets back a PENDING response
- Then load fails and FAILED status is cached
- Coordinator checks with the Historical if it has loaded the segment.
- Since we already have a cached FAILED status, we retry the load and send back a PENDING response (see snippet below)
- Repeat

https://github.com/apache/druid/blob/efb32810c48fd9c0bf81cefb15179f74a4cde661/server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java#L562-L563

**Comment by kfaraz** [comment]
Trying to handle all these case, I have made the following changes:
- Invalidate the cache if a different request type is received
- Rearrange and rename some methods to simplify the code
- Add javadocs where applicable
- Clean up `SegmentLoadDropHandlerTest` and add some more test cases

@gianm , let me know what you think.

**Comment by kfaraz** [comment]
There is one more case that requires handling, a concurrent load and drop of the same segment. I am currently working on this, will update the PR soon.

**Comment by gianm** [comment]
> concurrent load and drop of the same segment

When would this occur?

**Comment by gianm** [comment]
@kfaraz Thanks for outlining the scenario. IMO, it makes sense to include this fix, since it seems straightforward.

**Review by gianm — COMMENTED** [review]
Most of the changes LGTM. I had a couple of questions.

File: server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java [hunk 3]
```diff
@@ -272,21 +306,25 @@ private void loadSegment(DataSegment segment, DataSegmentChangeCallback callback
...
     final boolean loaded;
     try {
-      loaded = segmentManager.loadSegment(segment,
-              lazy,
-          () -> this.removeSegment(segment, DataSegmentChangeCallback.NOOP, false),
-              loadSegmentIntoPageCacheExec
+      loaded = segmentManager.loadSegment(
+          segment,
+          lazy,
+          () -> unannounceAndDropSegment(segment, DataSegmentChangeCallback.NOOP),
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java]: Yes, that is the intention here. But just to clarify, this was already the case and has not been introduced by this PR. I have only renamed the methods to remove ambiguity.

File: server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java [hunk 4]
```diff
@@ -90,19 +93,46 @@ public class SegmentLoadDropHandler implements DataSegmentChangeHandler
...
+   * <li>A duplicate request received at this point is immediately answered with PENDING.</li>
+   * <li>Once the load/drop finishes, the entry is updated to either SUCCESS or FAILED.</li>
+   * <li>A duplicate request received at this point is immediately answered with
+   * SUCCESS or FAILED and the entry is removed from the cache.</li>
+   * <li>If the first request itself finishes after the load or drop has already
+   * completed, it is answered with a SUCCESS or FAILED and the entry is removed
+   * from the cache.</li>
+   * </ol>
+   * <p>
+   * Maximum size of this cache must be significantly greater than the number of
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:118]: If the Coordinator sends a batch larger than the cache size, the Historical would not be able to have an entry for some of the requests in the cache. These requests would still be processed but their finished statuses would never be updated in the cache and thus never sent back to the Coordinator. 

The Coordinator would end up retrying these requests and the Historical might duplicate some work but there would never be any correctness issue, since we are not sending any response to the Coordinator for the un-cached requests.

Given that the default cache size is 100 and default request batch size is 1, we should be okay. We can update the docs to mention that the cache size must be at least 10x of the batch size.

[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:118]: Interestingly, `druid.segmentCache.statusQueueMaxSize` is not documented yet.

[the remark this answers is not in the document]
Comment (gianm) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:118]: Where does the 10x factor come from?

[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:118]: > Where does the 10x factor come from?

Just a safe-side measure in case Coordinator sends multiple batches of different requests. Not really based on any concrete thoughts.

[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:118]: Some more thoughts on this:
- The `Cache` can really just be a `Map`. The only advantage the `Cache` might offer is auto-expiry of requests that the Coordinator gave up on, but I don't think that is currently being used in the code anyway.
- Maybe we shouldn't even have a config for `batchSize`. The Coordinator would start by sending a single request to the historicals and in each HTTP response, each Historical could indicate how many requests it would like to take in the next batch. This could be based on the number of loading threads, number of pending requests.
- Given that the number of pending requests is always going to be small, we should just get rid of the `cacheSize` config and hard-code it to something like 1000. And if in any case, the Coordinator sends more requests than a Historical can handle, it immediately fails the request with an appropriate error message.

What do you think?

**Comment by rohangarg** [comment]
A couple of doubts : 

1. > DROP task does not do anything since the segment is not present in segmentsToDrop anymore. Thus the result holder remains stuck in pending state.

Are we sure that the DROP request actually gets stuck in pending state? It appears from the current code (https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java#L463) that the status is marked success after either the delete request is scheduled async or run inline.

2. Since we changed the return type object of the HTTP call, will it be backwards compatible? For instance if the historicals are updated before coordinator (as we suggest), could it happen that the historicals return new object types which an old coordinator wouldn't be able to understand?

File: server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gianm) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:138]: > The `Cache` can really just be a `Map`. The only advantage the `Cache` might offer is auto-expiry of requests that the Coordinator gave up on, but I don't think that is currently being used in the code anyway.

The Coordinator could crash and never follow up on a request, so it's still useful to expire stuff.

> Maybe we shouldn't even have a config for batchSize. The Coordinator would start by sending a single request to the historicals and in each HTTP response, each Historical could indicate how many requests it would like to take in the next batch. This could be based on the number of loading threads, number of pending requests.

Seems like a good idea! In addition, in case the Coordinator sends more commands than the Historical asks for, the Historical could reject the extra commands. (Although I am not sure if the protocol allows this.)

> Given that the number of pending requests is always going to be small, we should just get rid of the cacheSize config and hard-code it to something like 1000.

Sounds good to me. Fewer configs are good.

Btw, I am wondering now, what are the scenarios where the Historical might receive multiple copies of the same load or drop command? Is this something that happens in typical operation, or in some kind of abnormal case?

[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java:138]: > Btw, I am wondering now, what are the scenarios where the Historical might receive multiple copies of the same load or drop command?

Do you mean the same action on same segment or different actions on same segment? Both of these can actually happen under normal conditions:

#### Same action on same segment
Even under normal operation, the Coordinator keeps sending the same request to a Historical until the request finishes.
Essentially, the Coordinator sends a batch of requests to the Historical and waits for a response. Historical sends back a response as soon as any _one_ of the requests in this batch (or even a previously submitted batch which is still being tracked) finishes. So chances are that the Coordinator gets back a response with one request having status SUCCESS or FAILED and the others PENDING, because the Historical was still working on them when it completed the HTTP request. The Coordinator then re-sends the requests for which it had received a PENDING status. By then, the Historical might have already finished them and cached their results.

#### Different actions on same segment but after a delay
The case for a different action (i.e. load after drop) without ever having read the status of the previous action can happen during coordinator restarts or leader re-elections. Although, in these cases, it is not likely that the drop and load would happen concurrently.

#### Different actions on same segment happening concurrently
This is not expected to happen under normal conditions, and can only be caused by the Coordinator being in some weird state or a bug in the code.

**Comment by kfaraz** [comment]
> Are we sure that the DROP request actually gets stuck in pending state? It appears from the current code (https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java#L463) that the status is marked success after either the delete request is scheduled async or run inline.

Yes, @rohangarg , thanks for pointing this out. I have updated my comment. The DROP request is marked a success immediately after unannouncing the segment.

> Since we changed the return type object of the HTTP call, will it be backwards compatible? For instance if the historicals are updated before coordinator (as we suggest), could it happen that the historicals return new object types which an old coordinator wouldn't be able to understand?

The response payload has not changed, only the Java class name has changed. So the changes are backward compatible. Please let me know if you notice a discrepancy in this.

**Comment by gianm** [comment]
Removed from 27.0 in favor of #14670.

**Comment by kfaraz** [comment]
Closing this as this implementation is still buggy and the design requires some further thought.
I will create a fresh PR soon.
